Add options to checks

djjolicoeur opened the pull request:
@jcf , some things we were thinking about WRT this linter.  

1) Currently the newline is counted as one of the 80 chars in the line-limit.  I don't think I've seen this before and most IDE's (including mine) won't highlight on that whitespace.  I just trimmed the newline off the line count when we check. Alternatively, that could be toggled via an option passed in.

2) speaking of options, we wanted the ability to warn on certain conditions rather than hard error.  Ideally we would always keep things at 80 chars, but the hard limit makes for some odd formatting choices in some rare edge cases.  I threw together the ability to pass in some options and propagate warnings as well as errors with warnings not returning a non zero exit code.  I added the ability to all checks, but only implemented new options for the line lengths specifically.  This would give us the ability to evaluate lines slightly over the "soft limit" via the warning, and still have a hard stop for egregious line-length overruns.   

default behavior remains as is with the caveats that globs would be passed in via the `:path` option rather than the first arg to the linter and the 81st new line char not resulting in a non 0 exit status. 

This is thrown together rather quickly just to see if it was easily doable.  If the additions I have proposed here are agreeable, I can clean it up a bit.

djjolicoeur committed:
Pass args through to checkers,ability to warn,and line length warnings

Parse CLI args to map, update readme with proper project name

My linter wasn't passing the linter

jcf commented:
I have a much simpler version of this linter I've been working on in private, and I will be replacing the majority of the code in this library soon. For that reason I'm not planning on merging any significant changes.

In response to your points:

> Currently the newline is counted as one of the 80 chars in the line-limit. I don't think I've seen this before and most IDE's (including mine) won't highlight on that whitespace. I just trimmed the newline off the line count when we check. Alternatively, that could be toggled via an option passed in.

The newline shouldn't be included. A small PR that fixes that is welcome.

> speaking of options, we wanted the ability to warn on certain conditions rather than hard error. Ideally we would always keep things at 80 chars, but the hard limit makes for some odd formatting choices in some rare edge cases.

I'm not interested in making any of the rules optional. Maybe an option to configure the limit you want to enforce could be added, but as I'm happy with 80 columns everywhere I don't plan to make the line length logic more complicated/intelligent.

djjolicoeur commented:
@jcf  perhaps I was not clear enough in describing what was added in this PR.  I added the ability to pass options the the checkers that implement the rules, I did not make any rules optional. I added the option to add a warning to the line length rule in addition to a hard limit with the ability to configure both. To be clear, that means line length is configurable and you have the option to add a line length to warn on in addition to the line length.

If you do not like the idea of warnings in addition to errors (not instead of), I can remove the ability to do that and just keep the ability to pass in options like setting the line length through to the checkers. 

If you are planning significant changes in the near furture and this PR complicates that effort, that's fine.  We can always publish a jar from my fork to get us what we need in the meantime.
